### PR TITLE
Add Spell Checker to CI

### PR DESCRIPTION
**Pull request by h4shk4t**
## Which problem is this PR solving?
- Closes [#665](https://github.com/jaegertracing/documentation/issues/665)

## Description of the changes
- Simply added a new cspell job to the existing ci-test workflow.
- Will perform the spell check in the content directory and its subdirectories.
- Added the same set of dictionary as suggested in the reference PR in the issue thread. (Let me know if any more words need to be added)

### CODE REVIEW

**Review by yurishkuro — COMMENTED** [review]
doesn't look like it did anything in the CI run

File: .github/workflows/ci-test.yml [hunk 1]
```diff
@@ -29,3 +29,10 @@ jobs:
     - name: Build
       run: |
         make build
+    
+    - name: Spelling Check
+      uses: streetsidesoftware/cspell-action@v5.3.0
```
Comment (yurishkuro) [inline .github/workflows/ci-test.yml:34]: I don't like using magic GH actions for linters. It should be possible to run the linter locally, e.g. `make spellcheck`

Comment (h4shk4t) [reply .github/workflows/ci-test.yml:34]: Alright understood, I will make the changes and update again

File: .project-words.txt [hunk 2]
```diff
@@ -0,0 +1,143 @@
+accountingservice
```
[the remark this answers is not in the document]
Comment (h4shk4t) [reply .project-words.txt:1]: From the PR you shared on the GitHub issue thread, it seemed to have a good set of words which I thought would apply here as well as mentioned in the PR description. Should I manually add the words instead?

File: .github/workflows/ci-test.yml (hunk 1 above)
Comment (h4shk4t) [reply .github/workflows/ci-test.yml]: Wrote a python for spell check. Could you review?

File: .project-words.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (h4shk4t) [reply .project-words.txt:1]: I ran the script once and stored all the misspelled words in the first iteration as the words to ignore. Kindly check if they are sufficient. It is placed in scipts/utils folder.

File: scripts/utils/dictionary.txt [hunk 3]
```diff
@@ -0,0 +1,723 @@
+aspecto
```
Comment (yurishkuro) [inline scripts/utils/dictionary.txt:1]: this file doesn't make sense to me. E.g. why are commit hashes in there, or things like p50 or src? If the checker is so dumb as to flag every single string, then there's no value in it for us, it's always going to be too noisy.

Comment (h4shk4t) [reply scripts/utils/dictionary.txt:1]: Hmm, i get your point. Could you recommend a local spellchecker that runs well? I'll use that instead

Comment (yurishkuro) [reply scripts/utils/dictionary.txt:1]: I cannot, research is part of the task. I think OTEL website used one, other CNCF websites like Kubernetes might use one too.

Comment (h4shk4t) [reply scripts/utils/dictionary.txt:1]: So I researched a few spelling checker libraries and found 2 potential candidates. One is GNU Aspell and the other is [misspell](https://pkg.go.dev/github.com/client9/misspell/cmd/misspell) by client9. I tested both of them and I found Aspell to perform better at detecting wrong spellings and ignoring false positives like hexadecimal strings. This leaves us with 4 options (if we include the old options as well): Aspell, misspell, cspell (GitHub action), and pyspellchecker (in decreasing order of preference). I can continue looking for more spellcheckers if these are not satisfactory.

Comment (h4shk4t) [reply scripts/utils/dictionary.txt]: I wrote this implementation in Aspell. I created a dictionary on the basis of the initial run of Aspell over the contents. I tried my best to not accidentally add actual typos or misspellings generated from the command while writing the dictionary. One problem I encountered was of encoding of words like `Kröhling` and `Paixão`. Aspell was struggling to deal with the unicode characters of these words despite setting the encoding to UTF-8. The alternative was to use Hunspell, but it performs worse overall when it comes to checking spellings. Apart from that, the spellcheck seems to be working fine.

Comment (yurishkuro) [reply scripts/utils/dictionary.txt]: > words like Kröhling and Paixão

does it not work if we add these words to the dictionary?

File: my_dictionary.txt [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+autoscale
+autoscaler
+backend
+backends
+bak
+balancer
+bansal
+bohara
+bootcamp
+cassandra
```
Comment (yurishkuro) [inline my_dictionary.txt:27]: what kind of standard dictionary are you using? I would've expected these words to be pre-defined somewhere.

File: my_dictionary.txt [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+logz
+lookback
+loopback
+lyft
+macos
+mailgroup
+marck
+mentee
+mentees
+mentorship
```
Comment (yurishkuro) [inline my_dictionary.txt:150]: mentorship is not a recognized word?

File: my_dictionary.txt (hunk 4 above)
Comment (h4shk4t) [reply my_dictionary.txt:27]: I was not using a standard dictionary, what I did was run the spellchecker once and I manually added all the words from the output which were misclassified as incorrect spellings but were technical terms or proper nouns. Should I also add more words from another source?

File: my_dictionary.txt (hunk 5 above)
Comment (h4shk4t) [reply my_dictionary.txt:150]: Even I found that to be odd that Aspell's default dictionary doesnt recognise mentorship and some other words. However, when I tried alternatives like misspelling or hunspell I faced a lot more other issues like reporting hashes and hexadecimal strings as wrong spellings, etc. As Aspell has been the best in this regard I thought I'd just add it to the dictionary to make it work (there are not that many words with such false positives as far as I know)

Comment (yurishkuro) [reply my_dictionary.txt:150]: There are standard dictionaries in aspell ecosystem, aren't there? We need to use those and only add our dictionary with our specific terms.

Comment (h4shk4t) [reply my_dictionary.txt:150]: Yes this dictionary is on top of the standard dictionary offered by the aspell ecosystem.

File: scripts/utils/dictionary.txt (hunk 3 above)
Comment (h4shk4t) [reply scripts/utils/dictionary.txt]: On running the script I get the following error (after adding these words to the dictionary) 
image

**Comment by yurishkuro** [comment]
The spellchecker flags normal words, so I don't see it as a viable solution at this point. Maybe it needs better dictionaries. Also, I would argue it should not be checking test formatted as code, because that will always bring a lot of noise.

**Comment by h4shk4t** [comment]
@yurishkuro I did some digging and found the latest dictionary by `aspell` from 2020 in one of their mirrors. It shows some improvement where it is not flagging words like `mentorship`, `analytics` and more (verified this manually). However there are still a lot more common words which are still being flagged like `roadmap` and `onboarding`. I was trying Hunspell again, but I am facing issues with making it ignore hexadecimal values (will try to look for some bypasses to fix this issue in the meantime). Kindly advise if I should continue with `aspell` with a custom dictionary which also includes words like roadmap and onboarding, or to try and make `Hunspell` reduce false positives from code blocks and hexadecimal values.

**Comment by yurishkuro** [comment]
@h4shk4t I don't have preference in the tool, I only have preference for the end result - low friction :-)

**Comment by h4shk4t** [comment]
@yurishkuro Made the changes. Using `cspell` binary for spellchecking. Gives a promising accuracy and far lesser tolerance for false positives. Kindly let me know if this suffices. 

Best Regards

File: project-words.txt [hunk 6]
```diff
@@ -0,0 +1,166 @@
+hasparent
```
Comment (yurishkuro) [inline project-words.txt:1]: Please sort and add a linter check that it is sorted.

Is this list case sensitive?

File: scripts/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+pyspellchecker
+markdown-it-py
```
Comment (yurishkuro) [inline scripts/requirements.txt:2]: is this needed?

File: cspell.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (yurishkuro) [inline cspell.json:1]: adding a lot of files to the root dir, can we stash them all somewhere, like `scripts/cspell/*`?

File: .github/workflows/ci-test.yml [hunk 9]
```diff
@@ -28,4 +28,4 @@ jobs:
 
     - name: Build
       run: |
-        make build
+        make build
```
Comment (yurishkuro) [inline .github/workflows/ci-test.yml:31]: revert the change

File: project-words.txt (hunk 6 above)
Comment (h4shk4t) [reply project-words.txt]: Made the fixes

File: scripts/requirements.txt (hunk 7 above)
Comment (h4shk4t) [reply scripts/requirements.txt]: Made the fixes

File: .github/workflows/ci-test.yml (hunk 9 above)
Comment (h4shk4t) [reply .github/workflows/ci-test.yml]: Reverted the change

File: cspell.json (hunk 8 above)
Comment (h4shk4t) [reply cspell.json]: Made the fixes

File: .github/workflows/ci-test.yml [hunk 10]
```diff
@@ -27,5 +27,4 @@ jobs:
         hugo version
 
     - name: Build
-      run: |
-        make build
+      run: |
```
Comment (yurishkuro) [inline .github/workflows/ci-test.yml:30]: why do you need to change this file?

File: scripts/cspell/spellcheck.sh [hunk 11]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+if ! sort -c project-words.txt; then \
+    echo "project-words.txt is not sorted." && exit 1; \
+fi
+npm install cspell
+npm run spellcheck
```
Comment (yurishkuro) [inline scripts/cspell/spellcheck.sh:7]: in which directory will it run?

File: scripts/cspell/spellcheck.sh [hunk 12]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+if ! sort -c project-words.txt; then \
+    echo "project-words.txt is not sorted." && exit 1; \
+fi
```
Comment (yurishkuro) [inline scripts/cspell/spellcheck.sh:5]:
```suggestion
if ! sort -c project-words.txt; then
    echo "project-words.txt is not sorted."
    exit 1
fi
```

File: Makefile [hunk 13]
```diff
@@ -58,3 +58,6 @@ check-internal-links: clean build link-checker-setup run-link-checker
 
 check-all-links: clean build link-checker-setup
 	bin/htmltest --conf .htmltest.external.yml
+
+spellcheck:
```
Comment (yurishkuro) [inline Makefile:62]: this is not called from CI anywhere

File: .github/workflows/ci-test.yml (hunk 10 above)
Comment (h4shk4t) [reply .github/workflows/ci-test.yml:30]: I misunderstood what was asked. I reverted the change to the file by reverting back to the commit `Disable links checking...` on the ci-test.yml file.

File: scripts/cspell/spellcheck.sh (hunk 11 above)
Comment (h4shk4t) [reply scripts/cspell/spellcheck.sh:8]: It is being called from the `scripts/cspell/` directory. But the spellcheck script in `package.json` traverses to the content folder and performs a spellcheck on all the markdown files inside it.

File: scripts/cspell/spellcheck.sh (hunk 12 above)
Comment (h4shk4t) [reply scripts/cspell/spellcheck.sh]: Made the fixes

File: Makefile (hunk 13 above)
Comment (h4shk4t) [reply Makefile:62]: Now `ci-test.yml` is calling spellcheck and exits if a misspelling is found.

**Comment by yurishkuro** [comment]
The lint step is catching some real misspellings now, plus a few words that need to go to the dictionary

**Comment by h4shk4t** [comment]
> The lint step is catching some real misspellings now, plus a few words that need to go to the dictionary

Yes, I will add the new words to the dictionary (must have missed them earlier). What should I do about the misspellings?

**Review by yurishkuro — APPROVED** [review]
